horovod/common/operations.cc

```diff
@@ -1487,11 +1339,251 @@ void BackgroundThreadLoop(HorovodGlobalState& state) {
     (*it)(SHUT_DOWN_ERROR);
   }
 
-  MPI_Comm_free(&horovod_global.mpi_comm);
+  MPI_Comm_free(&state.mpi_comm);
+  MPI_Comm_free(&state.local_comm);
+  MPI_Comm_free(&state.cross_comm);
 
   MPI_Finalize();
 }
 
+// The coordinator currently follows a master-worker paradigm. Rank zero acts
+// as the master (the "coordinator"), whereas all other ranks are simply
+// workers. Each rank runs its own background thread which progresses in ticks.
+// In each tick, the following actions happen:
+//
+//      a) The workers send an MPIRequest to the coordinator, indicating what
+//      they would like to do (which tensor they would like to gather and
+//      reduce, as well as their shape and type). They repeat this for every
+//      tensor that they would like to operate on.
+//
+//      b) The workers send an empty "DONE" message to the coordinator to
+//      indicate that there are no more tensors they wish to operate on.
+//
+//      c) The coordinator receives the MPIRequests from the workers, as well
+//      as from its own TensorFlow ops, and stores them in a request table. The
+//      coordinator continues to receive MPIRequest messages until it has
+//      received MPI_SIZE number of empty "DONE" messages.
+//
+//      d) The coordinator finds all tensors that are ready to be reduced,
+//      gathered, or all operations that result in an error. For each of those,
+//      it sends an MPIResponse to all the workers. When no more MPIResponses
+//      are available, it sends a "DONE" response to the workers. If the process
+//      is being shutdown, it instead sends a "SHUTDOWN" response.
+//
+//      e) The workers listen for MPIResponse messages, processing each one by
+//      doing the required reduce or gather, until they receive a "DONE"
+//      response from the coordinator. At that point, the tick ends.
+//      If instead of "DONE" they receive "SHUTDOWN", they exit their background
+//      loop.
+bool RunLoopOnce(HorovodGlobalState& state, bool is_coordinator) {
+  // The coordinator sends a SHUTDOWN message to trigger shutdown.
+  bool should_shut_down = false;
+
+  // This delay determines thread frequency and MPI message latency
+  auto sleep_duration =
+      state.last_cycle_start +
+      std::chrono::microseconds(long(state.cycle_time * 1000.)) -
```

Comment:
Why is state.last_cyle_start in microseconds, but state.cycle_time in seconds?

Reply:
`state.cycle_time` is actually in milliseconds.  Sleep duration is not a number but a time point, and so it doesn't have fixed resolution.

Reply:
Ah, gotcha.  Minor thing for variable names, in general I think it's good to put units when they're otherwise implied.  So `cycle_time` -> `cycle_time_ms`.

Reply:
Good idea.